File: sdk/samples/all in on code/Visual Studio 2008/VBWinFormBindingNestedProperties/MyCustomTypeDescriptor.vb
'********************************* Module Header *********************************'
' Module Name:  MyCustomTypeDescriptor.vb
' Project:      VBWinFormBindToNestedProperties
' Copyright (c) Microsoft Corporation.
' 
' The sample demonstrates how to bind a DataGridView column to a nested property 
' in the data source.
' 
' This source is subject to the Microsoft Public License.
' See http://www.microsoft.com/opensource/licenses.mspx#Ms-PL.
' All other rights reserved.
' 
' THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
' EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED 
' WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
'\********************************************************************************'

Imports System.ComponentModel


Public Class MyCustomTypeDescriptor
    Inherits CustomTypeDescriptor

    Public Sub New(ByVal parent As ICustomTypeDescriptor)
        MyBase.New(parent)

    End Sub

    Public Overrides Function GetProperties(ByVal attributes() As Attribute) As PropertyDescriptorCollection
        ' Get original PropertyDescriptorCollection from the type
        Dim originalPds As PropertyDescriptorCollection = MyBase.GetProperties(attributes)
        Dim subPds As New List(Of PropertyDescriptor)

        Dim i As Integer = 0
        Do While (i < originalPds.Count)
            ' Get the chid properties of each original PropertyDescriptor
            Dim tempPds As PropertyDescriptorCollection = originalPds(i).GetChildProperties
            ' If the child properties' count is greater than 0, create a new 
            ' PropertyDescriptor for each of the child property and add it to the subPds
            If (tempPds.Count > 0) Then
                Dim j As Integer = 0
                Do While (j < tempPds.Count)
                    subPds.Add(New SubPropertyDescriptor(originalPds(i), tempPds(j), _
                        (originalPds(i).Name + ("_" + tempPds(j).Name))))
                    j = (j + 1)
                Loop
            End If
            i = (i + 1)
        Loop

        Dim array(originalPds.Count + subPds.Count) As PropertyDescriptor

        ' Copy the original PropertyDescriptorCollection to the array
        originalPds.CopyTo(array, 0)

        ' Copy all the PropertyDescriptor representing the child properties to the array
        subPds.CopyTo(array, originalPds.Count)
        Dim newPds As PropertyDescriptorCollection = New PropertyDescriptorCollection(array)

        ' Return the new PropertyDescriptorCollection containing both the PropertyDescriptor 
        ' for original properties and 
        ' child properties
        Return newPds
    End Function

    Public Overrides Function GetProperties() As PropertyDescriptorCollection
        ' Get original PropertyDescriptorCollection from the type
        Dim originalPds As PropertyDescriptorCollection = MyBase.GetProperties()
        Dim subPds As New List(Of PropertyDescriptor)

        Dim i As Integer = 0
        Do While (i < originalPds.Count)
            ' Get the chid properties of each original PropertyDescriptor
            Dim tempPds As PropertyDescriptorCollection = originalPds(i).GetChildProperties
            ' If the child properties' count is greater than 0, create a new PropertyDescriptor for
            ' each of the child property and add it to the subPds
            If (tempPds.Count > 0) Then
                Dim j As Integer = 0
                Do While (j < tempPds.Count)
                    subPds.Add(New SubPropertyDescriptor(originalPds(i), tempPds(j), _
                        (originalPds(i).Name + ("_" + tempPds(j).Name))))
                    j = (j + 1)
                Loop
            End If
            i = (i + 1)
        Loop

        Dim array(originalPds.Count + subPds.Count) As PropertyDescriptor

        ' Copy the original PropertyDescriptorCollection to the array
        originalPds.CopyTo(array, 0)

        ' Copy all the PropertyDescriptor representing the child properties to the array
        subPds.CopyTo(array, originalPds.Count)
        Dim newPds As PropertyDescriptorCollection = New PropertyDescriptorCollection(array)

        ' Return the new PropertyDescriptorCollection containing both the PropertyDescriptor 
        ' for original properties and child properties
        Return newPds
    End Function
End Class

Public Class MyTypeDescriptionProvider
    Inherits TypeDescriptionProvider

    Private td As ICustomTypeDescriptor

    Public Sub New()
        Me.New(TypeDescriptor.GetProvider(GetType(Person)))

    End Sub

    Public Sub New(ByVal parent As TypeDescriptionProvider)
        MyBase.New(parent)

    End Sub

    Public Overrides Function GetTypeDescriptor(ByVal objectType As Type, ByVal instance As Object) As ICustomTypeDescriptor
        If (td Is Nothing) Then
            td = MyBase.GetTypeDescriptor(objectType, instance)
            td = New MyCustomTypeDescriptor(td)
        End If
        Return td
    End Function
End Class

File: sdk/samples/all in on code/Visual Studio 2008/CSWinFormBindingNestedProperties/MyCustomTypeDescriptor.cs
/************************************* Module Header **************************************\
* Module Name:  MyCustomTypeDescriptor.cs
* Project:      CSWinFormBindToNestedProperties
* Copyright (c) Microsoft Corporation.
* 
* The sample demonstrates how to bind a DataGridView column to a nested property 
* in the data source.
* 
* This source is subject to the Microsoft Public License.
* See http://www.microsoft.com/opensource/licenses.mspx#Ms-PL.
* All other rights reserved.
* 
* THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
* EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
* WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
\******************************************************************************************/


using System;
using System.Collections.Generic;
using System.Text;
using System.ComponentModel;

namespace BindingNestedProperties
{
    public class MyCustomTypeDescriptor : CustomTypeDescriptor
    {
        public MyCustomTypeDescriptor(ICustomTypeDescriptor parent)

            : base(parent)
        {

        }

        public override PropertyDescriptorCollection GetProperties(Attribute[] attributes)
        {
            // get original PropertyDescriptorCollection from the type
            PropertyDescriptorCollection originalPds = base.GetProperties(attributes);
            List<PropertyDescriptor> subPds = new List<PropertyDescriptor>();

            for (int i = 0; i < originalPds.Count; i++)
            {
                // get the chid properties of each original PropertyDescriptor
                PropertyDescriptorCollection tempPds = originalPds[i].GetChildProperties();

                // if the child properties' count is greater than 0, create a new 
                // PropertyDescriptor for
                // each of the child property and add it to the subPds
                if (tempPds.Count > 0)
                {
                    for (int j = 0; j < tempPds.Count; j++)
                    {
                        subPds.Add(new SubPropertyDescriptor(originalPds[i], tempPds[j], 
                            originalPds[i].Name + "_" + tempPds[j].Name));
                    }
                }
            }

            PropertyDescriptor[] array = new PropertyDescriptor[originalPds.Count + subPds.Count];

            // copy the original PropertyDescriptorCollection to the array
            originalPds.CopyTo(array,0);

            // copy all the PropertyDescriptor representing the child properties to the array
            subPds.CopyTo(array, originalPds.Count);

            // create a new PropertyDescriptorCollection based on the array
            PropertyDescriptorCollection newPds = new PropertyDescriptorCollection(array);

            // return the new PropertyDescriptorCollection containing both the PropertyDescriptor 
            // for original properties and 
            // child properties
            return newPds;
        }

        public override PropertyDescriptorCollection GetProperties()
        {
            // get original PropertyDescriptorCollection from the type
            PropertyDescriptorCollection originalPds = base.GetProperties();
            List<PropertyDescriptor> subPds = new List<PropertyDescriptor>();

            for (int i = 0; i < originalPds.Count; i++)
            {
                // get the chid properties of each original PropertyDescriptor
                PropertyDescriptorCollection tempPds = originalPds[i].GetChildProperties();

                // if the child properties' count is greater than 0, create a new PropertyDescriptor for
                // each of the child property and add it to the subPds
                if (tempPds.Count > 0)
                {
                    for (int j = 0; j < tempPds.Count; j++)
                    {
                        subPds.Add(new SubPropertyDescriptor(originalPds[i], tempPds[j], 
                            originalPds[i].Name + "_" + tempPds[j].Name));
                    }
                }
            }

            PropertyDescriptor[] array = new PropertyDescriptor[originalPds.Count + subPds.Count];

            // copy the original PropertyDescriptorCollection to the array
            originalPds.CopyTo(array, 0);

            // copy all the PropertyDescriptor representing the child properties to the array
            subPds.CopyTo(array, originalPds.Count);

            // create a new PropertyDescriptorCollection based on the array
            PropertyDescriptorCollection newPds = new PropertyDescriptorCollection(array);

            // return the new PropertyDescriptorCollection containing both the PropertyDescriptor 
            // for original properties and 
            // child properties
            return newPds;

        }

    }

    public class MyTypeDescriptionProvider : TypeDescriptionProvider
    {
        private ICustomTypeDescriptor td;

        public MyTypeDescriptionProvider()
            : this(TypeDescriptor.GetProvider(typeof(Person)))
        {

        }

        public MyTypeDescriptionProvider(TypeDescriptionProvider parent)
            : base(parent)
        {

        }

        public override ICustomTypeDescriptor GetTypeDescriptor(Type objectType, object instance)
        {
            if (td == null)
            {
                td = base.GetTypeDescriptor(objectType, instance);
                td = new MyCustomTypeDescriptor(td);
            }
            return td;
        }

    }



}
